mobilecoind-dev-faucet/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+            req.set_monitor_id(monitor_id.clone());
+
+            let resp = mobilecoind_api_client
+                .get_public_address(&req)
+                .map_err(|err| format!("Failed getting public address: {}", err))?;
+
+            resp.b58_code
+        };
+
+        let monitor_printable_wrapper = PrintableWrapper::b58_decode(monitor_b58_address.clone())
+            .expect("Could not decode b58 address");
+        assert!(monitor_printable_wrapper.has_public_address());
+        let monitor_public_address = monitor_printable_wrapper.get_public_address();
+
+        // Get the network minimum fees and compute faucet amounts
+        let minimum_fees = {
+            let mut result = HashMap::<TokenId, u64>::default();
+
+            let resp = mobilecoind_api_client
+                .get_network_status(&Default::default())
+                .map_err(|err| format!("Failed getting network status: {}", err))?;
+
+            for (k, v) in resp.get_last_block_info().minimum_fees.iter() {
+                result.insert(k.into(), *v);
+            }
+
+            result
+        };
+
+        Ok((
+            monitor_id,
+            monitor_public_address.clone(),
+            monitor_b58_address,
+            minimum_fees,
+        ))
+    }
+
+    /// Handle a "post" to the faucet, which requests a payment from the faucet.
+    /// Returns either the mobilecoind success response or an error string.
+    pub async fn handle_post(
+        &self,
+        req: &JsonFaucetRequest,
+    ) -> Result<mc_mobilecoind_api::SubmitTxResponse, String> {
+        let printable_wrapper = PrintableWrapper::b58_decode(req.b58_address.clone())
+            .map_err(|err| format!("Could not decode b58 address: {}", err))?;
+
+        let public_address = if printable_wrapper.has_public_address() {
+            printable_wrapper.get_public_address()
+        } else {
+            return Err(format!(
+                "b58 address '{}' is not a public address",
+                req.b58_address
+            ));
+        };
+
+        let token_id = TokenId::from(req.token_id.unwrap_or_default().as_ref());
+
+        let utxo_record = self.worker.get_utxo(token_id)?;
+        log::trace!(
+            self.logger,
+            "Got a UTXO: key_image = {:?}, value = {}",
+            KeyImage::try_from(utxo_record.utxo.get_key_image()).unwrap(),
+            utxo_record.utxo.value
+        );
+
+        // Generate a Tx sending this specific TxOut, less fees
+        let mut req = mc_mobilecoind_api::GenerateTxFromTxOutListRequest::new();
+        req.set_account_key((&self.account_key).into());
+        req.set_input_list(vec![utxo_record.utxo].into());
+        req.set_receiver(public_address.clone());
+        req.set_token_id(*token_id);
+
+        let resp = self
+            .mobilecoind_api_client
+            .generate_tx_from_tx_out_list_async(&req)
+            .map_err(|err| format!("Failed to build Tx: {}", err))?
+            .await
```

Comment:
I might be wrong, but if I understand correctly whats happening here then it might be worth noting that the request handler is going to be executed from a tokio runner context, while this await is going to block on whatever runner grpcio is using. As such maybe its not worth using the `async` variant of the API client?

Reply:
No, each suspended Future tracks a waker to resume when it should be polled again, e.g. when data arrives (impls have to call the waker when `Future::poll` returns `Poll::Pending`). They can be running on different executors, including two Tokio runtimes. So the gRPC futures will likely be driven by the gRPC executor (in C) and then these handlers would use that data while running in a Tokio executor

Reply:
I had similar thoughts and I asked in the fog-core team standup (since there are a lot of async experts on that team now) if this is a bug, to my understanding, this isn't a bug and this is the way we're supposed to do it.

Here's busyjay on the issue: https://github.com/tikv/grpc-rs/issues/520

* tokio and grpcio have different thread pools and different executors in this program, the grpcio one is based in some C code.
* calls to `tokio::task::spawn` (or any tokio function that calls that) must be made in the context of the tokio executor, tokio will panic like `grpc-poll-0' panicked at 'not currently running on the Tokio runtime.` (https://github.com/tikv/grpc-rs/issues/338)
* awaiting a grpcio future in a tokio context is not bad though, if a tokio task goes to sleep on a grpcio future, i think the grpcio executor will ultimately call the "waker" from tokio via a generic and compatible API, and this is expected to work
* using an async API here is good because in theory this thread can be used by tokio to handle more requests
* even if we switched to calling a sync API, it is likely that the sync API works by calling the async api and then using `.await` (https://docs.rs/grpcio/latest/src/grpcio/client.rs.html#41)

that last point may be a good argument for switching to calling the sync version again, since it is calling await under the hood for us anyways, but we would have less code / less error handling boilerplate and still reap the benefits of async

Reply:
> * using an async API here is good because in theory this thread can be used by tokio to handle more requests

That's precisely the benefit of async APIs vs blocking.

> * even if we switched to calling a sync API, it is likely that the sync API works by calling the async api and then using `.await` (https://docs.rs/grpcio/latest/src/grpcio/client.rs.html#41)
> 
> that last point may be a good argument for switching to calling the sync version again, since it is calling await under the hood for us anyways, but we would have less code / less error handling boilerplate and still reap the benefits of async

Note that they're calling `block_on` to block the current thread, which is very different from `.await`